File: 13.roman-to-integer.go
import "fmt"

/*
 * @lc app=leetcode id=13 lang=golang
 *
 * [13] Roman to Integer
 */

// @lc code=start
func romanToInt(s string) int {
	var answer int = 0
	var greatest int = 0
	c := map[rune]int{
		'I': 1,
		'V': 5,
		'X': 10,
		'L': 50,
		'C': 100,
		'D': 500,
		'M': 1000,
	}

	for i := len(s) - 1; i >= 0; i-- {
		letter := s[i]
		number := c[rune(letter)]
		if number >= greatest {
			greatest = number
			fmt.Println(greatest)
			answer += number
			continue
		}

		answer = answer - number
	}
	return answer
}

// @lc code=end

File: 13.roman-to-integer.cpp
/*
 * @lc app=leetcode id=13 lang=cpp
 *
 * [13] Roman to Integer
 */

// @lc code=start
class Solution {
public:
    int romanToInt(string s) {
        int answer = 0;
        int greatest = 0;

        unordered_map<char, int> hmap;
        hmap.insert({'I', 1});
        hmap.insert({'V', 5});
        hmap.insert({'X', 10});
        hmap.insert({'L', 50});
        hmap.insert({'C', 100});
        hmap.insert({'D', 500});
        hmap.insert({'M', 1000});

        for (int i = s.size() - 1; i >= 0; i--) {
            char c = s[i];
            int val = hmap[c];
            if (val >= greatest) {
                greatest = val;
                cout << greatest << endl;
                answer += val;
                continue;
            }
            answer -= val;
        }
        return answer;
    }
};
// @lc code=end

